plugins/linux-capture/pipewire.c

```diff
@@ -0,0 +1,1110 @@
+	.process = on_process_cb,
+};
+
+static void on_core_error_cb(void *user_data, uint32_t id, int seq, int res,
+			     const char *message)
+{
+	UNUSED_PARAMETER(seq);
+
+	obs_pipewire_data *obs_pw = user_data;
+
+	blog(LOG_ERROR, "[pipewire] Error id:%u seq:%d res:%d (%s): %s", id,
+	     seq, res, g_strerror(res), message);
+
+	pw_thread_loop_signal(obs_pw->thread_loop, FALSE);
+}
+
+static void on_core_done_cb(void *user_data, uint32_t id, int seq)
+{
+	UNUSED_PARAMETER(seq);
+
+	obs_pipewire_data *obs_pw = user_data;
+
+	if (id == PW_ID_CORE)
+		pw_thread_loop_signal(obs_pw->thread_loop, FALSE);
+}
+
+static const struct pw_core_events core_events = {
+	PW_VERSION_CORE_EVENTS,
+	.done = on_core_done_cb,
+	.error = on_core_error_cb,
+};
+
+static void play_pipewire_stream(obs_pipewire_data *obs_pw)
+{
+	struct spa_pod_builder pod_builder;
+	const struct spa_pod *params[1];
+	uint8_t params_buffer[1024];
+	struct obs_video_info ovi;
+
+	obs_pw->thread_loop = pw_thread_loop_new("PipeWire thread loop", NULL);
+	obs_pw->context = pw_context_new(
+		pw_thread_loop_get_loop(obs_pw->thread_loop), NULL, 0);
+
+	if (pw_thread_loop_start(obs_pw->thread_loop) < 0) {
+		blog(LOG_WARNING, "Error starting threaded mainloop");
+		return;
+	}
+
+	pw_thread_loop_lock(obs_pw->thread_loop);
+
+	/* Core */
+	obs_pw->core = pw_context_connect_fd(
+		obs_pw->context, fcntl(obs_pw->pipewire_fd, F_DUPFD_CLOEXEC, 3),
+		NULL, 0);
+	if (!obs_pw->core) {
+		blog(LOG_WARNING, "Error creating PipeWire core: %m");
+		pw_thread_loop_unlock(obs_pw->thread_loop);
+		return;
+	}
+
+	pw_core_add_listener(obs_pw->core, &obs_pw->core_listener, &core_events,
+			     obs_pw);
+
+	/* Stream */
+	obs_pw->stream = pw_stream_new(
+		obs_pw->core, "OBS Studio",
+		pw_properties_new(PW_KEY_MEDIA_TYPE, "Video",
+				  PW_KEY_MEDIA_CATEGORY, "Capture",
+				  PW_KEY_MEDIA_ROLE, "Screen", NULL));
+	pw_stream_add_listener(obs_pw->stream, &obs_pw->stream_listener,
+			       &stream_events, obs_pw);
+
+	/* Stream parameters */
+	pod_builder =
+		SPA_POD_BUILDER_INIT(params_buffer, sizeof(params_buffer));
+
+	obs_get_video_info(&ovi);
+	params[0] = spa_pod_builder_add_object(
+		&pod_builder, SPA_TYPE_OBJECT_Format, SPA_PARAM_EnumFormat,
+		SPA_FORMAT_mediaType, SPA_POD_Id(SPA_MEDIA_TYPE_video),
+		SPA_FORMAT_mediaSubtype, SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw),
+		SPA_FORMAT_VIDEO_format,
+		SPA_POD_CHOICE_ENUM_Id(
+			4, SPA_VIDEO_FORMAT_RGBA, SPA_VIDEO_FORMAT_RGBx,
+			SPA_VIDEO_FORMAT_BGRx, SPA_VIDEO_FORMAT_BGRA),
+		SPA_FORMAT_VIDEO_size,
+		SPA_POD_CHOICE_RANGE_Rectangle(&SPA_RECTANGLE(320, 240),
+					       &SPA_RECTANGLE(1, 1),
+					       &SPA_RECTANGLE(4096, 4096)),
+		SPA_FORMAT_VIDEO_framerate,
+		SPA_POD_CHOICE_RANGE_Fraction(
+			&SPA_FRACTION(ovi.fps_num, ovi.fps_den),
+			&SPA_FRACTION(0, 1), &SPA_FRACTION(144, 1)));
+
+	pw_stream_connect(
+		obs_pw->stream, PW_DIRECTION_INPUT, obs_pw->pipewire_node,
+		PW_STREAM_FLAG_AUTOCONNECT | PW_STREAM_FLAG_MAP_BUFFERS, params,
```

Comment:
Do we really want to map buffers? We should never have to read the buffers on the cpu unless they were already on the cpu.

Reply:
PipeWire only really maps memory buffers (either memfd or raw buffers). DMA-BUF fds are not mapped. This is merely a convenience for us not have to map / unmap memory buffers manually.